base/compiler/utilities.jl

```diff
@@ -112,6 +112,11 @@ function retrieve_code_info(linfo::MethodInstance)
         return get_staged(linfo)
     else
         # TODO: post-inference see if we can swap back to the original arrays?
+        println(linfo)
+        println(linfo.def)
+        println(typeof(linfo.def))
+        println(typeof(m.source))
+        println(m.source)
```

Comment:
With a `make debug`, this debugging output produces lots of lines like (nb I've manually shortened the arrays)
```julia
print(Core.CoreSTDOUT, Type{Core.MethodInstance})
Core.print(...)
Method
Array{UInt8, 1}
Array{UInt8, (66,)}[0x00, 0x03, 0x00, 0x00, 0x00, ...]
show(Core.CoreSTDOUT, Type{Core.MethodInstance})
Core.show(...)
Method
Array{UInt8, 1}
Array{UInt8, (119,)}[0x00, 0x03, 0x00, 0x00, ...]
```
which all seem fine.

Then we get to the fatal error
```julia
validate_code_in_debug_mode(Core.MethodInstance, Core.CodeInfo, String)
Core.Compiler.validate_code_in_debug_mode(...)
Method
Core.MethodInstance
Array{UInt8, (452,)}[0x00, 0x07, 0x00, 0x00, ...]
error during bootstrap:
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "compiler/bootstrap.jl" line 8: MethodError(f=typeof(Core.Compiler.copy_code_info)(), args=(Array{UInt8, (452,)}[0x00, 0x07, 0x00, 0x00, ...],), world=0x0000000000000e5d)))
```
Notice that `m.source` is a `MethodInstance`, [which shouldn't happen](https://github.com/JuliaLang/julia/blob/39e5742557cf56f26e1eccd28ab52d258ae19863/src/julia.h#L277) and is inconsistent with the actual value that is printed. Interestingly, you can explain this if this is an off-by-one bug in the field offset and you're picking up `unspecialized` instead. Since I added a field to `method_t`, this seems plausible. Know of any places where field offsets are likely to be hard-coded?

Reply:
Got it: https://github.com/JuliaLang/julia/blob/2346be896129f594ce2e6104f612828468e63bfc/src/jltypes.c#L2218
Just happened to stumble across it while searching.